src/hotspot/cpu/ppc/downcallLinker_ppc.cpp

```diff
@@ -35,6 +105,255 @@ RuntimeStub* DowncallLinker::make_downcall_stub(BasicType* signature,
                                                 bool needs_return_buffer,
                                                 int captured_state_mask,
                                                 bool needs_transition) {
-  Unimplemented();
-  return nullptr;
+  int code_size = native_invoker_code_base_size + (num_args * native_invoker_size_per_arg);
+  int locs_size = 1; // must be non-zero
+  CodeBuffer code("nep_invoker_blob", code_size, locs_size);
+  DowncallStubGenerator g(&code, signature, num_args, ret_bt, abi,
+                          input_registers, output_registers,
+                          needs_return_buffer, captured_state_mask,
+                          needs_transition);
+  g.generate();
+  code.log_section_sizes("nep_invoker_blob");
+
+  RuntimeStub* stub =
+    RuntimeStub::new_runtime_stub("nep_invoker_blob",
+                                  &code,
+                                  g.frame_complete(),
+                                  g.framesize(),
+                                  g.oop_maps(), false);
+
+#ifndef PRODUCT
+  LogTarget(Trace, foreign, downcall) lt;
+  if (lt.is_enabled()) {
+    ResourceMark rm;
+    LogStream ls(lt);
+    stub->print_on(&ls);
+  }
+#endif
+
+  return stub;
+}
+
+void DowncallStubGenerator::generate() {
+  Register callerSP            = R2, // C/C++ uses R2 as TOC, but we can reuse it here
+           tmp                 = R11_scratch1, // same as shuffle_reg
+           call_target_address = R12_scratch2; // same as _abi._scratch2 (ABIv2 requires this reg!)
+  VMStorage shuffle_reg = _abi._scratch1;
+  JavaCallingConvention in_conv;
+  NativeCallingConvention out_conv(_input_registers);
+  ArgumentShuffle arg_shuffle(_signature, _num_args, _signature, _num_args, &in_conv, &out_conv, shuffle_reg);
+
+#ifndef PRODUCT
+  LogTarget(Trace, foreign, downcall) lt;
+  if (lt.is_enabled()) {
+    ResourceMark rm;
+    LogStream ls(lt);
+    arg_shuffle.print_on(&ls);
+  }
+#endif
+
+  // Stack frame size computation:
+  // We use the number of input VMStorage elements because PPC64 requires slots for all arguments
+  // (even if they are passed in registers), at least 8.
+  // This may be a bit more than needed when HFA is used (see CallArranger.java).
+  // (native_abi_reg_args is native_abi_minframe plus space for 8 argument register spill slots)
+  assert(_abi._shadow_space_bytes == frame::native_abi_minframe_size, "expected space according to ABI");
+  // The Parameter Save Area needs to be at least 8 slots for ABIv1.
+  // ABIv2 allows omitting it when all parameters can get passed in registers. We currently don't optimize this.
+  // For ABIv2, we only need (_input_registers.length() > 8) ? _input_registers.length() : 0
```

Comment:
The PSA is also needed if the parameter list is variable in length. Is the expression `(_input_registers.length() > 8) ? _input_registers.length() : 0` correct in that case too?
Otherwise: `ABIv2 allows omitting it if the caller's prototype indicates that stack parameters are not expected. We currently don't optimize this.`